After setting level in runtime, line number(%l) doesn't work

Hi. I've installed latest version, v6.4.3.

If logging level is set at runtime, the variable for line number (%l) doesn't work.

### Code

```javascript
'use strict';

const log4js = require('log4js');

const opts = {
  appenders: {
    customized: {
      type: 'stdout',
      layout: {
        type: 'pattern',
        pattern: '%[[%d] [%5p] [%c](%3l)%] %m',
      },
    },
  },
  categories: {
    default: {
      appenders: ['customized'],
      level: 'debug',
      enableCallStack: true,
    },
  },
};

const logger = log4js.configure(opts).getLogger('TEST');

logger.error('before set level');
logger.level = 'info';
logger.error('after set level');
```

### RESULT

```
[2022-03-17T12:22:41.695] [ERROR] [TEST]( 26) before set level
[2022-03-17T12:22:41.698] [ERROR] [TEST](   ) after set level
```

Note to self: 
It seems that setting `logger.level` will reset `logger.useCallStack`.

```javascript
> logger.error('before set level');
[2022-03-17T23:56:28.242] [ERROR] [TEST](  1) before set level
undefined

> logger.useCallStack;
true

> logger.level = 'info';
'info'

> logger.useCallStack;
false

> logger.error('after set level');
[2022-03-17T23:56:28.564] [ERROR] [TEST](   ) after set level
undefined

> logger.useCallStack = true
true

> logger.error('after set useCallStack');
[2022-03-17T23:57:40.910] [ERROR] [TEST](  1) after set useCallStack
undefined
```

--- a/lib/categories.js
+++ b/lib/categories.js
@@ -141,6 +141,9 @@ configuration.addListener((config) => {
 
 const setup = (config) => {
   categories.clear();
+  if (!config) {
+    return;
+  }
 
   const categoryNames = Object.keys(config.categories);
   categoryNames.forEach((name) => {
@@ -162,7 +165,7 @@ const setup = (config) => {
 };
 
 const init = () => {
-  setup({ categories: { default: { appenders: ['out'], level: 'OFF' } } });
+  setup();
 };
 init();
 
@@ -174,29 +177,27 @@ const configForCategory = (category) => {
     debug(`configForCategory: ${category} exists in config, returning it`);
     return categories.get(category);
   }
+
+  let sourceCategoryConfig;
   if (category.indexOf('.') > 0) {
-    debug(`configForCategory: ${category} has hierarchy, searching for parents`);
-    return configForCategory(category.substring(0, category.lastIndexOf('.')));
+    debug(`configForCategory: ${category} has hierarchy, cloning from parents`);
+    sourceCategoryConfig = { ...configForCategory(category.substring(0, category.lastIndexOf('.'))) };
+  } else {
+    if (!categories.has('default')) {
+      setup({ categories: { default: { appenders: ['out'], level: 'OFF' } } });
+    }
+    debug('configForCategory: cloning default category');
+    sourceCategoryConfig = { ...categories.get('default') };
   }
-  debug('configForCategory: returning config for default category');
-  return configForCategory('default');
+  categories.set(category, sourceCategoryConfig);
+  return sourceCategoryConfig;
 };
 
 const appendersForCategory = category => configForCategory(category).appenders;
-const getLevelForCategory = category => configForCategory(category).level;
 
+const getLevelForCategory = category => configForCategory(category).level;
 const setLevelForCategory = (category, level) => {
-  let categoryConfig = categories.get(category);
-  debug(`setLevelForCategory: found ${categoryConfig} for ${category}`);
-  if (!categoryConfig) {
-    const sourceCategoryConfig = configForCategory(category);
-    debug('setLevelForCategory: no config found for category, '
-      + `found ${sourceCategoryConfig} for parents of ${category}`);
-    categoryConfig = { appenders: sourceCategoryConfig.appenders };
-    categoryConfig.enableCallStack = sourceCategoryConfig.enableCallStack;
-  }
-  categoryConfig.level = level;
-  categories.set(category, categoryConfig);
+  configForCategory(category).level = level;
 };
 
 const getEnableCallStackForCategory = category => configForCategory(category).enableCallStack === true;

--- a/lib/categories.js
+++ b/lib/categories.js
@@ -193,6 +193,7 @@ const setLevelForCategory = (category, level) => {
     debug('setLevelForCategory: no config found for category, '
       + `found ${sourceCategoryConfig} for parents of ${category}`);
     categoryConfig = { appenders: sourceCategoryConfig.appenders };
+    categoryConfig.enableCallStack = sourceCategoryConfig.enableCallStack;
   }
   categoryConfig.level = level;
   categories.set(category, categoryConfig);

--- a/lib/logger.js
+++ b/lib/logger.js
@@ -60,7 +60,7 @@ class Logger {
   get level() {
     return levels.getLevel(
       categories.getLevelForCategory(this.category),
-      levels.TRACE
+      levels.OFF
     );
   }
 

--- a/test/tap/logger-test.js
+++ b/test/tap/logger-test.js
@@ -3,12 +3,14 @@ const debug = require("debug")("log4js:test.logger");
 const sandbox = require("@log4js-node/sandboxed-module");
 const callsites = require("callsites");
 const levels = require("../../lib/levels");
+const categories = require("../../lib/categories");
 
 const events = [];
 const messages = [];
 const Logger = sandbox.require("../../lib/logger", {
   requires: {
     "./levels": levels,
+    "./categories": categories,
     "./clustering": {
       isMaster: () => true,
       onlyOnMaster: fn => fn(),
@@ -268,6 +270,44 @@ test("../../lib/logger", batch => {
     t.equal(events[0].lineNumber, 15);
     t.equal(events[0].columnNumber, 25);
     t.equal(events[0].callStack, "test callstack");
+
+    t.end();
+  });
+
+  batch.test("creating/cloning of category", t => {
+    const defaultLogger = new Logger("default");
+    defaultLogger.level = "trace";
+    defaultLogger.useCallStack = true;
+
+    t.test("category should be cloned from parent/default if does not exist", assert => {
+      const originalLength = categories.size;
+
+      const logger = new Logger("cheese1");
+      assert.equal(categories.size, originalLength + 1, "category should be cloned");
+      assert.equal(logger.level, levels.TRACE, "should inherit level=TRACE from default-category");
+      assert.equal(logger.useCallStack, true, "should inherit useCallStack=true from default-category");
+      assert.end();
+    });
+
+    t.test("changing level should not impact default-category or useCallStack", assert => {
+      const logger = new Logger("cheese2");
+      logger.level = "debug";
+      assert.equal(logger.level, levels.DEBUG, "should be changed to level=DEBUG");
+      assert.equal(defaultLogger.level, levels.TRACE, "default-category should remain as level=TRACE");
+      assert.equal(logger.useCallStack, true, "should remain as useCallStack=true");
+      assert.equal(defaultLogger.useCallStack, true, "default-category should remain as useCallStack=true");
+      assert.end();
+    });
+
+    t.test("changing useCallStack should not impact default-category or level", assert => {
+      const logger = new Logger("cheese3");
+      logger.useCallStack = false;
+      assert.equal(logger.useCallStack, false, "should be changed to useCallStack=false");
+      assert.equal(defaultLogger.useCallStack, true, "default-category should remain as useCallStack=true");
+      assert.equal(logger.level, levels.TRACE, "should remain as level=TRACE");
+      assert.equal(defaultLogger.level, levels.TRACE, "default-category should remain as level=TRACE");
+      assert.end();
+    });
 
     t.end();
   });

